net/libfab/libfab.c

```diff
@@ -0,0 +1,2829 @@
+		struct m0_net_buffer *nb;
+
+		m0_tl_for(m0_net_tm, &net->ntm_q[i], nb) {
+			if (nb->nb_timeout < now) {
+				nb->nb_flags |= M0_NET_BUF_TIMED_OUT;
+				libfab_buf_done(nb->nb_xprt_private,
+						-ETIMEDOUT);
+			}
+		} m0_tl_endfor;
+	}
+	M0_POST(libfab_tm_invariant(ftm));
+}
+
+/**
+ * Finds pending buffers completions and completes them.
+ *
+ * A buffer is placed on tm::ftm_done queue when its operation is done, but the
+ * completion call-back cannot be immediately invoked, for example, because
+ * completion happened in a synchronous context.
+ */
+static void libfab_tm_buf_done(struct m0_fab__tm *ftm)
+{
+	struct m0_fab__buf *buffer;
+	int                 nr = 0;
+
+	M0_PRE(libfab_tm_is_locked(ftm) && libfab_tm_invariant(ftm));
+	m0_tl_for(fab_buf, &ftm->ftm_done, buffer) {
+		fab_buf_tlist_del(buffer);
+		libfab_buf_complete(buffer, buffer->fb_status);
+		nr++;
+	} m0_tl_endfor;
+
+	if (nr > 0 && ftm->ftm_ntm->ntm_callback_counter == 0)
+		m0_chan_broadcast(&ftm->ftm_ntm->ntm_chan);
+	M0_POST(libfab_tm_invariant(ftm));
+}
+
+/**
+ * Used to handle incoming connection request events
+ */
+static uint32_t libfab_handle_connect_request_events(struct m0_fab__tm *tm)
+{
+	struct m0_fab__ep        *ep = NULL;
+	struct m0_fab__conn_data *cd;
+	struct m0_fab__ep_name    en;
+	struct fid_eq            *eq;
+	struct fi_eq_err_entry    eq_err;
+	struct fi_eq_cm_entry     entry;
+	uint32_t                  event;
+	int                       rc;
+
+	eq = tm->ftm_pep->fep_listen->pep_res.fpr_eq;
+	rc = fi_eq_read(eq, &event, &entry,
+			(sizeof(entry) + sizeof(struct m0_fab__conn_data)), 0);
+	if (rc > 0 && rc >= sizeof(entry)) {
+		if (event == FI_CONNREQ) {
+			memset(&en, 0, sizeof(en));
+			cd = (struct m0_fab__conn_data*)entry.data;
+			libfab_ep_ntop(cd->fcd_netaddr, &en);
+			libfab_fab_ep_find(tm, &en, cd->fcd_straddr, &ep);
+			rc = libfab_conn_accept(ep, tm, entry.info);
+			if (rc != FI_SUCCESS)
+				M0_LOG(M0_ERROR, "Conn accept failed %d", rc);
+			fi_freeinfo(entry.info);
+		}
+	} else if (rc == -FI_EAVAIL) {
+		memset(&eq_err, 0, sizeof(eq_err));
+		rc = fi_eq_readerr(eq, &eq_err, 0);
+		if (rc != sizeof(eq_err)) {
+			M0_LOG(M0_ERROR, "fi_eq_readerr returns error =%s",
+			       fi_strerror((int) -(rc)));
+		} else {
+			M0_LOG(M0_ERROR, "fi_eq_readerr provider err no %d:%s",
+				eq_err.prov_errno,
+				fi_eq_strerror(eq, eq_err.prov_errno,
+					       eq_err.err_data, NULL, 0));
+		}
+	}
+	return 0;
+}
+
+/**
+ * Check connetion established and shutdown events for transmit endpoint
+ */
+static void libfab_txep_event_check(struct m0_fab__ep *txep,
+				    struct m0_fab__active_ep *aep,
+				    struct m0_fab__tm *tm)
+{
+	int rc;
+
+	rc = libfab_check_for_event(aep->aep_tx_res.ftr_eq);
+	if (rc == FI_CONNECTED) {
+		libfab_pending_bufs_send(txep);
+		aep->aep_tx_state = FAB_CONNECTED;
+	} else if (rc == FI_SHUTDOWN) {
+		/* Reset and reopen endpoint */
+		libfab_txep_init(aep, tm);
+	}
```

Comment:
Is it possible to have other types of events?
If yes, they are ignored?

Please comments to explain that.

Reply:
Added comments.
All other events are discarded.